app/models/course/achievement.rb

```diff
@@ -45,4 +45,27 @@ def permitted_for!(course_user)
   def precluded_for!(course_user)
     course_users.delete(course_user) if course_users.exists?(course_user.id)
   end
+
+  def initialize_duplicate(duplicator, other)
+    self.course = duplicator.duplicate(other.course)
+    duplicate_badge(other) if other.badge_url
+    self.achievement_conditions = duplicator.duplicate(other.achievement_conditions)
+    # Duplicate actable object directly and let the acting_as gem create the Condition object
+    self.conditions = duplicator.duplicate(other.conditions.map(&:actable)).map(&:acting_as)
+  end
+
+  private
+
+  # Helper function to figure out the storage mechanism and duplicate the badge
+  # Adapted from https://github.com/equivalent/copy_carrierwave_file
+  #
+  # @param other [Course::Achievement] the source achievement object to be duplicated
+  def duplicate_badge(other)
+    case other.badge.send(:storage).class.name
+    when 'CarrierWave::Storage::File'
+      self.badge = File.new(File.join(Rails.root, '/public' + other.badge.url))
```

Comment:
Think `other.badge.path` will do, the `public` folder is configurable, thus might not be `/public`